packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,384 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""HTTP connection, channel, server, and handler"""
+
+import asyncio
+import json
+import logging
+import threading
+from asyncio import CancelledError
+from http.server import BaseHTTPRequestHandler, HTTPServer
+from typing import Optional, Set, cast
+from urllib.parse import parse_qs, urlparse
+from uuid import uuid4
+
+import yaml
+
+from aea.configurations.base import ConnectionConfig, PublicId
+from aea.connections.base import Connection
+from aea.mail.base import Address, Envelope, EnvelopeContext, URI
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPChannel:
+    """A wrapper for an RESTful API with an internal HTTPServer."""
+
+    def __init__(
+        self,
+        address: Address,
+        host: str,
+        port: int,
+        api_spec: str,
+        connection_id: PublicId,
+        excluded_protocols: Optional[Set[PublicId]] = None,
+    ):
+        """
+        Initialize a channel.
+
+        :param address: the address
+        """
+        self.address = address
+        self.host = host
+        self.port = port
+        self.connection_id = connection_id
+        self.in_queue = None  # type: Optional[asyncio.Queue]
+        self.loop = None  # type: Optional[asyncio.AbstractEventLoop]
+        self.excluded_protocols = excluded_protocols
+        # self.thread = Thread(target=self.receiving_loop)
+        self.lock = threading.Lock()
+        self.stopped = True
+
+        if api_spec is not None:
+            self._process_api(api_spec)
+        logger.info(f"Initialized the HTTP Server on HOST: {host} and PORT: {port}")
+
+    def _process_api(self, api_spec):
```

Comment:
Here I'd consume the api_specification_path as an argument and construct a helper object which lets me verify any given path

Reply:
To start, I'll create a **PetStoreHelper** class within connection.py that emulates an "echo" handler. It takes an init Envelope from _process_api() that specifies what are the valid path and stores in an internal dict. It then uses that to verify incoming requests and returns the appropriate response.

Reply:
I hope it is fine **APIVerificationHelper** class it nested under HTTPConnection. Next step, it may probably be its own Skill handler that receives and responds both in Envelope format through the Multiplexer.